File: node_js_ds_and_algos_implementations/heaps/kth_largest_element_in_stream.js
/**
 * @param {number} k
 * @param {number[]} nums
 */
/*
Approach:
The idea is that we have a nums list of length N, but we only want to get 
the Kth largest elements in that list of nums. 
For example:
[4, 5, 8, 2]
k = 3
we only want the 3 largest elements in this list of 4, the kth largest element
would be 4 here

In the constructor, we want to use a min heap instead of a max heap. The concept is reverse thinking,
the smallest element is at the front, and the largest in the back for a min heap. Therefore,
by adding all elements and then popping until the minHeap.size() === k, we get the kth largest elements

min heap: [2, 4, 5, 8]
size is 4, k = 3, so we need to pop() one
min heap is now [4, 5, 8]

Now if we want to add an element:
num to add = 3
min heap : [4, 5, 8]
after adding:
[3,4,5,8]

Now the min heap is now size 4, so we need to pop from min heap to get it back to size k
[4,5,8]

we now return the front element, which is the kth largest element (4)

now if we add 5
min heap: [4, 5, 8], after adding: [4, 5, 5, 8]

pop from min heap to get back to size k
[5, 5, 8],

we return the front element, which is the kth largest element (5) 

Time Complexity:
Constructor: O(NLogN), which is the creation of the min heap when enqueueing 
Add: O(LogN), since we need at most one enqueue operation (OLogN to bubble up any possible smaller elements) and one pop operation which is LogN
*/
var KthLargest = function(k, nums) {
    this.k = k
    this.minHeap = new MinPriorityQueue()
    for (let n of nums){
        this.minHeap.enqueue(n)
    }
    while (this.minHeap.size() > this.k){
        this.minHeap.dequeue()
    }
};

/** 
 * @param {number} val
 * @return {number}
 */
KthLargest.prototype.add = function(val) {
    /*
    Note that the heap automatically stores each value as an object
    like so:
    {priority: val, element: val},
    so you need to access the element attribute to get the number
    */
    this.minHeap.enqueue(val)
    if (this.minHeap.size() > this.k){
        this.minHeap.dequeue()
    }
    return this.minHeap.front().element
};

/** 
 * Your KthLargest object will be instantiated and called as such:
 * var obj = new KthLargest(k, nums)
 * var param_1 = obj.add(val)
 */

File: neetcode all and misc questions/Heaps/kth_largest_element_in_stream.py
"""
https://leetcode.com/problems/kth-largest-element-in-a-stream/
Time: O(NLogN + MLogK), where M is the amount of calls to "add"
NLogN is the time taken to convert the array to a heap
Space: O(N)

Concept: 

Note that when you print a heap out in python, it's in the "heap" represented order (if you were to create a tree out of it),
but not necessarily in the array's sorted order, in a similar way to an array representation of an inorder binary tree traversal.


1) Use Min Heap
kth largest element in the list example:
k = 3
heap = [1, 2, 3, 5, 6]
Because k is 3, the third largest element in this list is 3

2) Because we only need k, we can pop any elements off that are less than k, so 1 and 2 in this case

heap = [3, 5, 6]

3) When adding new elements, we do heap push, and then immediately pop off the smallest element

heap = [3, 5, 4, 6]

After heappop, the 4 gets bubbled up to the root as it's now the min element

heap = [4, 5, 6]

"""
class KthLargest:
    import heapq
    def __init__(self, k: int, nums: List[int]):
        self.k = k
        self.heap = nums
        heapq.heapify(self.heap)
        """
        Revisited on 10/1/2024,
        Time Complexity:
        O(NLogN) heapify + O(MLogN), where M is the amount of calls made to self.add()
        put all items of the stream into the min heap so they are sorted, but then
        pop out until only k items are left.
        The idea is that the kth largest element is always the first element of the min heap, 
        kind of like a "sliding window", where the heap only contains the largest k elements
        """
        for i in range(len(nums)-k):
            heapq.heappop(self.heap)

    def add(self, val: int) -> int:
        """ 
        push val to the heap. If the heap is now greater than k, pop out the first value. 
        Return the first value of the heap. 
        This also works if we add a value that is less than the top of the heap,
        since we'd add it the top, and then immediately remove it
        """
        heapq.heappush(self.heap, val)
        if len(self.heap) > self.k:
            heapq.heappop(self.heap)
        return self.heap[0]
        
import heapq
class KthLargest:
    
    def __init__(self, k: int, nums: List[int]):
        self.k = k
        self.nums = nums
        heapq.heapify(self.nums)
        # use a min heap, and pop out until only k items are left
        # in the heap
        while len(self.nums) > self.k:
            heapq.heappop(self.nums)
        
    def add(self, val: int) -> int:
        heapq.heappush(self.nums, val)
        # if we've exceeded k amount, pop
        if len(self.nums) > self.k:
            heapq.heappop(self.nums)
        # the first item in the list will always be the kth largest 
        return self.nums[0]


# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)